### PR TITLE
Fix bug where remote suspension causes local instance to remove remote follows

### PR DESCRIPTION
**Pull request by ShadowJonathan**
This does not have an issue number, but I noticed this bug while looking at the comment.

The comment presumes that this code will only be run when the suspension comes locally, but I've observed the code being triggered when a remote account gets suspended remotely, causing the local server to issue follow rejects to that account.

According to the comment, this is not intended, and any (temporary) local suspension will effectively remove remote follows.

### CODE REVIEW

**Comment by ShadowJonathan** [comment]
The bug happens like follows;
- Alice@Amazing.com follows Bart@Beautiful.net
- Server A suspends Alice@Amazing.com
- Server A distributes this suspension via account update, triggering [this code](https://github.com/mastodon/mastodon/blob/6ad203337496ed8c74ac6eaa3f4a9943afd3c50a/app/services/suspend_account_service.rb#L43)
- Server B gets account update, triggering [this service](https://github.com/mastodon/mastodon/blob/cf6f70799b128d385d830d1f55eae8046bea3c60/app/services/activitypub/process_account_service.rb#L3)
- The Service [sets suspension](https://github.com/mastodon/mastodon/blob/cf6f70799b128d385d830d1f55eae8046bea3c60/app/services/activitypub/process_account_service.rb#L146) to remote, and then [proceeds](https://github.com/mastodon/mastodon/blob/cf6f70799b128d385d830d1f55eae8046bea3c60/app/services/activitypub/process_account_service.rb#L55) to [process the suspension change](https://github.com/mastodon/mastodon/blob/cf6f70799b128d385d830d1f55eae8046bea3c60/app/services/activitypub/process_account_service.rb#L168)
- It enters `SuspensionWorker`, which calls `SuspendAccountService`, which calls `reject_remote_follows!`, which was supposed to only run under the assumption of local suspensions.

Ergo, suspending a local account gets remote servers to send `Reject` activities to that account's follows.

**Comment by ClearlyClaire** [comment]
This indeed sounds like a bug worth fixing, but I don't think this should be done before ensuring users are offered to and effectively able to sever relationships with suspended users. This is not currently possible from the web interface.

**Comment by ShadowJonathan** [comment]
Ah, you want to fix the frontend first before this?

**Comment by ClearlyClaire** [comment]
> Ah, you want to fix the frontend first before this?

Yes! Also, I need to check whether that's just a front-end thing or if the backend prevents from severing the relationships.

**Comment by ShadowJonathan** [comment]
I think it was just never necessary, as this code is always triggered on every suspension, so it just simply wasn't ever possible to still be following a suspended user, if the user would get a "this user is suspended" message.

I haven't seen any code which prevents following/unfollowing a suspended user, but I missed this code first time when looking for what happens upon suspension, so....

**Comment by ClearlyClaire** [comment]
Looking into it the backend does not prevent unfollowing or blocking suspended accounts, which is a good thing. However, it refuses returning relationships with suspended accounts (see `app/controllers/api/v1/accounts/relationships_controller.rb`).

This means the UI can't know you are following them or they are following you, and thus cannot present you with the relevant options. This is an easy change, but this will require further change to the Web UI and perhaps other clients as well to properly handle suspended users with relationships without e.g. prompting you to try to follow a suspended account.

**Comment by ShadowJonathan** [comment]
@ClearlyClaire with #27667 merged, could this PR proceed?

**Review by ClearlyClaire — DISMISSED** [review]

**Comment by ShadowJonathan** [comment]
@renchap could this PR get some extra eyes? We got bitten by this again today

**Comment by renchap** [comment]
@ClearlyClaire approved the PR so I guess this is fine, I will let her merge it.

**Comment by ClearlyClaire** [comment]
Sorry, dismissed my review because I need to think more about what happens when you're followed by some remote suspended user. Merging this would cause remote suspended users to still get your posts, both public and followers-only.

**Comment by mjankowski** [comment]
Does the recent introduction of the severed relationship stuff make this one any more/less reviewable at this point?

**Comment by mjankowski** [comment]
Might need rebase, and CI here is frozen.

Also waiting on contemplation of remote suspended follows interaction w/ the change here.

**Comment by ShadowJonathan** [comment]
Will take a look, thanks for the bump

**Comment by ClearlyClaire** [comment]
> Does the recent introduction of the severed relationship stuff make this one any more/less reviewable at this point?

In this case, the suspension is the remote server suspending a user of theirs, so severed relationships notifications do not apply.

> Also waiting on contemplation of remote suspended follows interaction w/ the change here.

Consider the following situation: `bob@a` from server A follows `alice@b` from server B. `bob@a` gets suspended on A. The current behavior is that upon receiving the suspension notice, B breaks the follow relationship from `bob@a` to `alice@b`. This PR changes the behavior so that B does not break the follow relationship, and will still deliver posts (both public and private) to them.

This is not necessarily a big issue, but one could imagine an attacker following someone then “pretending” to be suspended while still reading incoming posts. Now, thanks to #27667 this should not be as much of an issue as it previously was, but an attacker could still go unnoticed, especially with apps that may not have taken advantage of the API changes in #27667.

Maybe it would make sense to alter the delivery logic to skip delivering new posts to suspended users.

**Comment by ClearlyClaire** [comment]
> Maybe it would make sense to alter the delivery logic to skip delivering new posts to suspended users.

This seems to be already covered by #27509, which I had forgotten about.

**Comment by ClearlyClaire** [comment]
> Closing on assumption that the other (now merged) PR handles this. Please reopen if there's more to extract here.

No, there has been a misunderstanding, the thing I was saying is that the PR I was liking to addressed one of my concerns with the current PR.

**Comment by mjankowski** [comment]
Ah, sorry! -- is this good as-is? Reviewable as-is? Need any rebase or more contemplation after that other one?

**Comment by ClearlyClaire** [comment]
The other PR was missing a bit (it only skipped suspended accounts for indirectly reached accounts, not followers), so this will have to be addressed first, but at this time I don't think the current PR needs any more change. Need to think about it a bit more, but I think the prerequisite are almost in place.

**Comment by ShadowJonathan** [comment]
Rebased on `main`

**Comment by ClearlyClaire** [comment]
> The other PR was missing a bit (it only skipped suspended accounts for indirectly reached accounts, not followers), so this will have to be addressed first, but at this time I don't think the current PR needs any more change. Need to think about it a bit more, but I think the prerequisite are almost in place.

#33030 has addressed that missing bit, so I *think* this PR should be good to go, but I need to give it another careful look.

**Comment by ShadowJonathan** [comment]
Rebased again (on `main`), @ClearlyClaire @mjankowski is there anything else I can do to push this PR along? Or what is this realistically blocked on?

**Comment by ShadowJonathan** [comment]
(I'm unsure why one wholly unrelated test failed, retrigger please?)

**Comment by mjankowski** [comment]
> Rebased again (on `main`)

Fun!

> is there anything else I can do to push this PR along?

It'd be nice to add some spec coverage showing that follows for a remote-suspension-account get preserved after this change?

> Or what is this realistically blocked on?

Not sure.

> (I'm unsure why one wholly unrelated test failed, retrigger please?)

This is a known intermittent failure (various discussions elsewhere on how to resolve). I can't retrigger, but this is def not related to your changes.

**Review by ClearlyClaire — APPROVED** [review]
Before this PR, when a remote account is suspended by their own server, their local followers are removed to avoid disclosing private information.

The concerns are twofold:
- users being unable to (soft-)block remotely-suspended followers. This concern has been addressed by #27667 (though mobile apps may or may not handle that)
- users' private posts still being sent to remote suspended users. This concern has been addressed by #33030

**Pull request merged**
